VTK 7 (py2 and py3)

Korijn opened the pull request:
We built VTK 7 successfully using this recipe for the full range of Python versions (2.6, 2.7, 3.3, 3.4, 3.5) on x64 windows and linux architectures, so I reckon it should work on x86 and OS X as well.

The main reason for writing it is that the VTK package in the official anaconda channels is lagging behind and doesn't support python 3.x.

Looking forward to improving this recipe together with you all.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/vtk```) and found some lint.

Here's what I've got...


For **recipes/vtk**:

 * When defining a source/url please add a sha256, sha1 or md5 checksum (sha256 preferably).

pelson commented:
Hi @Korijn! Thanks for submitting!

> The main reason for writing it is that the VTK package in the official anaconda channels is lagging behind and doesn't support python 3.x.

Awesome. Glad conda-forge can help!

> Looking forward to improving this recipe together with you all.

👍 

The CI looks a bit scary when it is all red like that, but hopefully we can iterate on it. To break down the problems by OS:

Linux:

```
-- Found X11: /usr/lib64/libX11.so
CMake Error at Rendering/OpenGL2/CMakeLists.txt:177 (message):
  X11_Xt_LIB could not be found.  Required for VTK X lib.
```

I'm afraid I know very little about X11, but I notice that libxt isn't part of the manylinux spec. For now, you can add a ```yum_requirements.txt``` containing ```libXt-devel``` in your recipe directory to get the necessary build dependency (see https://github.com/conda-forge/staged-recipes/blob/master/scripts/run_docker_build.sh#L45 for how this has been implemented).

He is currently away, but @jakirkham may be able to help in deciding whether we need to package libxt within conda-forge, or whether we simply make it a requirement for using the conda-forge distribution.

------

OSX:

```
+ mkdir build
+ cd build
++ uname
+ '[' Darwin == Linux ']'
+ make -j2
make: *** No targets specified and no makefile found.  Stop.
Command failed: /bin/bash -x -e /Users/travis/build/conda-forge/staged-recipes/recipes/vtk/build.sh
```

----

Windows. Took 43min on Py27, 1 hr (and timed out) on py34 x86.

The timeouts are problematic - we have an hour to build on item of the matrix, and unfortunately that is a hard limit put onto us from AppVeyor. I don't know that there is anything we can do about that in terms of either a) what we build or b) asking for longer.

The win x64 (py27):

```
C:\Miniconda-x64\conda-bld\work\VTK-7.0.0>REM ===== end generated header ===== 
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:29 (project):
  No CMAKE_C_COMPILER could be found.
 
CMake Error at CMakeLists.txt:29 (project):
  No CMAKE_CXX_COMPILER could be found.

-- Configuring incomplete, errors occurred!
See also "C:/Miniconda-x64/conda-bld/work/VTK-7.0.0/build/CMakeFiles/CMakeOutput.log".
See also "C:/Miniconda-x64/conda-bld/work/VTK-7.0.0/build/CMakeFiles/CMakeError.log".
 
Microsoft (R) Visual C++ Express Edition Version 9.0.30729.1.
Copyright (C) Microsoft Corp 2007. All rights reserved.
 
The following files were specified on the command line:
 
	VTK.sln
 
These files could not be found and will not be loaded. 
Microsoft (R) Visual C++ Express Edition Version 9.0.30729.1.
Copyright (C) Microsoft Corp 2007. All rights reserved.
 
The following files were specified on the command line:
 
	VTK.sln
 
These files could not be found and will not be loaded. 
Microsoft (R) Visual C++ Express Edition Version 9.0.30729.1.
Copyright (C) Microsoft Corp 2007. All rights reserved.
 
The following files were specified on the command line:
 
	VTK.sln
 
These files could not be found and will not be loaded. 
Microsoft (R) Visual C++ Express Edition Version 9.0.30729.1.
Copyright (C) Microsoft Corp 2007. All rights reserved.
 
The following files were specified on the command line:
 
	VTK.sln
 
These files could not be found and will not be loaded.Command failed: C:\Windows\system32\cmd.exe /c call bld.bat
```


Hope the failures aren't demoralizing - this is a really cool recipe to package and I'm super pleased you've submitted a PR to fix the current state of VTK on conda.`

Korijn commented:
I can address most of these problems. About the x64 2.7 build; do you have a custom identifier for the c compiler? This recipe is tested and working on a Windows 7 VM with VS2008 Pro SP1 which has the x64 compiler built-in.

msarahan commented:
I think the tricky part is that we have AppVeyor here.  It uses VS 2008
Express, which has no x64 support.  Instead, we use the VS Compiler for
Python 2.7, which is the same compiler, but no SDKs.  That's what Conda
1.20.1 fixed primarily - the search order.  Unfortunately, it brought in
other issues.

On Mon, Apr 25, 2016 at 5:30 PM Korijn van Golen <notifications@github.com>
wrote:

> I can address most of these problems. About the x64 2.7 build; do you have
> a custom identifier for the c compiler? This recipe is tested and working
> on a Windows 7 VM with VS2008 Pro SP1 which has the x64 compiler built-in.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/conda-forge/staged-recipes/pull/453#issuecomment-214550630>
>

jakirkham commented:
So, I guess it is time to bring back the VS 2008 64-bit workaround ( https://github.com/conda-forge/staged-recipes/pull/251 ). That didn't last long.

> Unfortunately, it brought in other issues.

I take it you are aware of this issue ( https://github.com/conda-forge/staged-recipes/issues/448 ) then. Did that move upstream? Do we have any idea of what is causing it.

msarahan commented:
I can't make sense of it right away.  https://ci.appveyor.com/project/conda-forge/xonsh-feedstock/build/1.0.14/job/827pntj23q0xxt0p#L202 seems to be one issue - Windows 7.1 SDK not having the 64-bit compiler, possibly.  I'm pretty sure @patricksnape tested that, though.

The other issue appears to be the same: https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.1375/job/jq2rrc7uwjb8iehw#L213

This should be a quick fix once we figure out what the correct value should be.  I can help push a quick release.

PS: @patricksnape it is better to set variables like:

    set "VARNAME=value"

rather than ```set VARNAME="value"```

    https://ci.appveyor.com/project/conda-forge/xonsh-feedstock/build/1.0.14/job/827pntj23q0xxt0p#L198

Both treat spaces well, but the former doesn't have quotes in the end variable value, and is easier to deal with.

Sorry I didn't catch this on the PR.

jakirkham commented:
Might need `freeglut`. Take a look at this comment ( https://github.com/conda-forge/staged-recipes/pull/285#issuecomment-210678237 ).

jakirkham commented:
To help preserve our sanity, I moved the weird Windows build issue into this issue ( https://github.com/conda/conda-build/issues/895 ). IMHO we have at least confirmed there is a conda-build bug at this point. Maybe we can focus the discussion there. Tracking it over multiple PRs, issues, and feedstocks is getting a little out of hand.

jakirkham commented:
Also, I'm wrong about needing that workaround. I forgot the new issue is restricted to Windows 64-bit Python 3.4. So, we shouldn't need the old VS 2008 64-bit workaround as the downgrade of `conda-build` is not needed on Python 2.7.

jakirkham commented:
Ignore the above. Also, don't dig any deeper. I know the cause.

jakirkham commented:
So, this PR ( https://github.com/conda-forge/staged-recipes/pull/455 ) should fix things for you here at staged-recipes. Want to give it a try?

pelson commented:
> So, this PR ( #455 ) should fix things for you here at staged-recipes. Want to give it a try?

That is, just rebase this PR against upstream/master and get #455 pulled in.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/vtk```) and found it was in an excellent condition.

Korijn commented:
Changes:

* Rebased against upstream/master
* Added sha1 checksum
* Dropped `mesa` requirement in favor of `freeglut`
* Added `yum_requirements.txt` for `libXi-devel`
* Added `.gitattributes` to guard newlines in `build.sh`

jakirkham commented on recipes/vtk/bld.bat:
```diff
@@ -0,0 +1,33 @@
+cmake --build . --target INSTALL --config %BUILD_CONFIG%
+if errorlevel 1 exit 1
+
+exit /b 0
```
Please add a terminal newline.

jakirkham commented on recipes/vtk/.gitattributes:
```diff
@@ -0,0 +1,3 @@
+# Set the default behavior, in case people don't have core.autocrlf set.
+*.bash text=lf
+*.sh text=lf
```
Should this be included?

Korijn replied:
I work on a Windows machine, and it kept converting newlines. This made my life a little bit easier.

jakirkham replied:
Understand. Could you maybe use a global gitattributes file?

My concern is that we really don't want to include this sort of stuff in a recipe. If we do decide we want this, it should really belong in the repo level not just this recipe. Also, we will want to add it to conda-smithy.

If you think this is generally useful, could you please add this to a new PR and move this to a top level directory so we can have this discussion separately?

jakirkham commented on recipes/vtk/meta.yaml:
```diff
@@ -0,0 +1,55 @@
+
+about:
+  home: http://www.vtk.org/
+  license: BSD
```
Version?

Korijn replied:
Will add it.

jakirkham commented on recipes/vtk/meta.yaml:
```diff
@@ -0,0 +1,55 @@
+
+requirements:
+  build:
+    - gcc                              # [linux and (py27 or py34 or py35)]
```
Why do we need `gcc` here? If we don't need it, can we please drop it?

Korijn replied:
We previously compiled using the system `gcc` but it would cause linking problems at runtime. The problems went away when we made sure to use the `gcc` provided by conda.

jakirkham replied:
Could you please share what linking problems were encountered?

ivoflipse replied:
We built on a dev's machine, so hard to tell what the exact cause was. Given that these builds run on CI servers, we should be able to rely on the systems gcc version since the other packages do as well

jakirkham commented on recipes/vtk/meta.yaml:
```diff
@@ -0,0 +1,55 @@
+    - python
+  run:
+    - freeglut                             # [linux and (py27 or py34 or py35)]
+    - libgcc                           # [linux and (py27 or py34 or py35)]
```
Same story. Would like to drop this too.

jakirkham replied:
Could we please also fix the selector indentation?

jakirkham commented:
Is it possible to disable X11 support on Mac? If so, that's what I would do here.

Korijn commented:
I'm clueless about X11 and also about Mac. There's probably a cmake flag to disable it.

jakirkham commented:
There is. Just set this on Mac only `-DVTK_USE_X:BOOL=OFF`.

jakirkham commented:
The thing with Mac is they used to have X11 support baked in, but they decided to drop it. Though one can still install it and Apple still supports this open source effort. It is just something we can't expect to find on the user's system. Maybe one day we can add a package to install it in a conda environment. Until then, disabling is the right way to go.

jakirkham commented:
So, I don't think he mentioned it, but @msarahan had a recipe for this that was in the process of being added. Not to speak for him, but you seem motivated to get this in and I think we are happy to see someone pushing this forward so please continue (plus I'm sure @msarahan has tons on his plate and doesn't mind someone taking this off it 😉). Though you might be able to pick up a few useful things from his recipe, which can be found in this PR ( https://github.com/conda-forge/staged-recipes/pull/285 ).

jakirkham commented:
@stuarteberg, given your use of VTK, maybe you have some useful feedback here. Please feel free to share.

jakirkham commented on recipes/vtk/build.sh:
```diff
@@ -0,0 +1,41 @@
+        -DPYTHON_LIBRARY:FILEPATH=$library_file_path
+fi
+
+make -j${CPU_COUNT}
```
Could we add `ctest`. Also, you may need [this]( https://github.com/conda-forge/eccodes-feedstock/issues/1#issuecomment-215695989 ).

jakirkham replied:
We have been a little cautious about using `CPU_COUNT` as the CIs don't always report this correctly. For instance, they may report the number of CPUs for the node you are on instead of noting how many of these you get. Could we try removing this? If the build time is too long, we can play with this more.

jakirkham commented on recipes/vtk/bld.bat:
```diff
@@ -0,0 +1,33 @@
+
+cmake -G"%GENERATOR_NAME%" -Wno-dev -DINSTALL_LIB_DIR:PATH="%LIBRARY_LIB%" -DPYTHON_INCLUDE_DIR:PATH="%PREFIX%/include" -DVTK_WRAP_PYTHON:BOOL="1" -DCMAKE_CONFIGURATION_TYPES:STRING="Debug;Release;MinSizeRel;RelWithDebInfo" -DINSTALL_BIN_DIR:PATH="%LIBRARY_BIN%" -DVTK_PYTHON_VERSION:STRING="%PY_VER%" -DVTK_INSTALL_PYTHON_MODULE_DIR:PATH="%SP_DIR%" -DPYTHON_LIBRARY:FILEPATH="%PYTHON_LIBRARY%" -DIN ...
+
+cmake --build . --target ALL_BUILD --config %BUILD_CONFIG%
```
Could we also add `ctest` here?

Korijn replied:
I've never worked with `ctest` before, do you have any examples?

stuarteberg commented:
>given your use of VTK, maybe you have some useful feedback here.

I don't have much to offer.  I use a pretty old version of VTK.  Recipe is here:
https://github.com/ilastik/ilastik-build-conda/blob/master/vtk/build.sh

One possible suggestion: Our recipe sets the `-DVTK_USE_SYSTEM_ZLIB`/`PNG`/`JPEG`/`TIFF`/`LIBXML` flags, just to make sure VTK is using the same version of those libraries that the rest of our conda packages are using.  Not sure if that's a big deal.

jakirkham commented:
> Our recipe sets the `-DVTK_USE_SYSTEM_ZLIB`/`PNG`/`JPEG`/`TIFF`/`LIBXML` flags, just to make sure VTK is using the same version of those libraries that the rest of our conda packages are using.

👍 Love this idea. @Korijn, would you mind adding this please?

jakirkham commented:
> I use a pretty old version of VTK.

We can support multiple versions if that is of interest to you, @stuarteberg.

grlee77 commented:
Thanks for working on this.

Just comparing CMAKE flags used here vs. those in #285, are any of the following flags needed/desired here?  Maybe you do not need VTK_PYTHON_VERSION if you are already explicitly setting the various library and include paths?

```
-DVTK_HAS_FEENABLEEXCEPT=OFF
-DModule_vtkRenderingMatplotlib=ON
-DVTK_PYTHON_VERSION=${PY_VER:0:1}
```
I don't know if Mayavi uses that Matplotlib module or not, but if so it would be good to enable it.  I think PARAVIEW uses it to render equations.

Also, you can probably add maintainers from that PR to this one to have more help going forward.

Korijn commented:
I went ahead and processed some of the simpler changes, will dive into the nitty gritty with `bld.bat` and `build.sh` later. Like @msarahan the amount of time I have available is unpredictable.

jakirkham commented on recipes/vtk/meta.yaml:
```diff
@@ -0,0 +1,62 @@
+
+requirements:
+  build:
+    - freeglut                             # [linux and (py27 or py34 or py35)]
```
These are the only Python's we build for so the selector can just be `# [linux]`.

jakirkham commented on recipes/vtk/meta.yaml:
```diff
@@ -0,0 +1,62 @@
+    - cmake
+    - python
+  run:
+    - freeglut                             # [linux and (py27 or py34 or py35)]
```
Same as above.

jakirkham commented on recipes/vtk/yum_requirements.txt:
```diff
@@ -0,0 +1 @@
+libXt-devel
```
Need `mesa-libGLU-devel` also.

jakirkham commented on recipes/vtk/build.sh:
```diff
@@ -0,0 +1,41 @@
+        -DCMAKE_INSTALL_PREFIX="${PREFIX}" \
+        -DCMAKE_INSTALL_RPATH:STRING="${PREFIX}/lib" \
+        -DBUILD_DOCUMENTATION=OFF \
+        -DBUILD_TESTING=OFF \
```
Guessing we need this enabled to run the tests.

jakirkham commented on recipes/vtk/build.sh:
```diff
@@ -0,0 +1,41 @@
+
+if [ `uname` == Linux ]; then
+    CC=${PREFIX}/bin/gcc
+    CXX=${PREFIX}/bin/g++
```
These should be dropped there is a system compiler we can sue for this.

Korijn replied:
We had issues earlier where linking errors would occur due to the system compiler producing builds incompatible with Anaconda's python.exe. Why would this not be an issue on conda-forge?

jakirkham replied:
Not sure I understand how they were incompatible. We regularly build things with the system compiler that link against Python. Could we please try doing so?

jakirkham commented on recipes/vtk/build.sh:
```diff
@@ -0,0 +1,41 @@
+
+    cmake .. \
+        -DCMAKE_C_COMPILER=$CC \
+        -DCMAKE_CXX_COMPILER=$CXX \
```
Again this should just pickup the system compiler so we shouldn't even need to specify.

jakirkham commented on recipes/vtk/build.sh:
```diff
@@ -0,0 +1,41 @@
+        -DPYTHON_INCLUDE_PATH:PATH=$include_path \
+        -DPYTHON_LIBRARY:FILEPATH=$library_file_path
+fi
+
```
Don't we need `cmake` here? Otherwise how do we build for Mac?

EelcoHoogendoorn replied:
We're trying to do so, but it's not entirely clear to us how that works on Linux.

Korijn replied:
On second thought, `make install` should work just fine on both platforms I think...

ivoflipse commented on recipes/vtk/bld.bat:
```diff
@@ -0,0 +1,49 @@
+REM tell cmake where Python is
+set PYTHON_LIBRARY=%PREFIX%\libs\python%PY_VER:~0,1%%PY_VER:~2,1%.lib
+
+cmake .. -G"%GENERATOR_NAME%" ^
```
We could use NMake here too, if it fails on 2.7

Korijn commented:
We've rebased again, processed the comments, and cleaned up using the things we learned from the `u3d` recipe. Adding tests will be challenging and probably cause timeouts on the builds... but we'll see!

ivoflipse commented:
@thewtex or @dlonie Would any of you guys at Kitware mind having a look at our build scripts to see what we can do to get our VTK recipe working?

Korijn commented:
The good news:

* Windows and Linux build scripts work for Python 2k and 3k.
* I've built and installed both locally and run some test scripts to start a renderer; works!
* OS X seems to be a simple fix, we're currently pointing to the python library incorrectly.

The bad news:

* Timeouts are a problem, even with all optional components disabled:
  * Setting `BUILD_TESTING:BOOL=ON` triggers the downloading of external test data which increases build time with about 1 hour.
  * The Windows build takes about 2,5 hours.

I'm currently working under a tight schedule and need to get at least the Linux packages out there. I propose skipping Windows and OS X for now, merging, and opening a PR where we can continue to work on the build timeout issues on CI, integrating proper testing and possibly supporting features that e.g. Mayavi needs. @jakirkham, thoughts?

ocefpaf commented:
> I'm currently working under a tight schedule and need to get at least the Linux packages out there. I would like to propose the following:
- skip Windows: not even 1 configuration will complete before the build timeout
- skip OS X: we don't have access to OS X systems to work on the build script
- restrict Linux to x64 py2.7 and x64 py.35: it should complete for these 2 important configurations before the build timeout

@Korijn that is a sound strategy (I did something similar with `gdal` some time ago).

:+1:

Korijn commented:
@ocefpaf cool. I commited the proposed changes, let's see if CI turns all-green.

ocefpaf commented:
Finger crossed!

Korijn commented on recipes/vtk/build.sh:
```diff
@@ -0,0 +1,39 @@
+    -DCMAKE_BUILD_TYPE=$BUILD_CONFIG \
+    -DCMAKE_INSTALL_PREFIX:PATH="${PREFIX}" \
+    -DCMAKE_INSTALL_RPATH:PATH="${PREFIX}/lib" \
+    -DINSTALL_PKGCONFIG_DIR:PATH=$PKG_CONFIG_PATH \
```
`INSTALL_PKGCONFIG_DIR` is unused according to a CMake warning. will remove that in future PR.

Korijn commented:
@ocefpaf can we merge? I'll open a PR on the feedstock directly afterwards.

ocefpaf commented:
Thanks @Korijn! I know the pain to fight against the CI limitations. Thanks for leading this effort!

jakirkham commented:
> * Timeouts are a problem, even with all optional components disabled:
  * Setting `BUILD_TESTING:BOOL=ON` triggers the downloading of external test data which increases build time with about 1 hour.
  * The Windows build takes about 2,5 hours per configuration.
  * The Linux build takes about 50 minutes per configuration.
  * As we want to build for x86/x64, py2.7, py3.4 and py3.5 (10 configurations in total), there is no way we'll be able to stay under the timeout limits of 1 hour on Windows (AppVeyor), 2 hours on Linux (CircleCI) and 50 minutes on OS X (Travis CI).

We had the same rough experience adding OpenCV, @Korijn. This sounds totally sensible to me. I'm glad this was added as is.

Also, good job getting this to build. 🎉

thewtex commented:
CC: @demarle @cryos @vibraphone

Korijn commented:
This work is continued here:

https://github.com/conda-forge/vtk-feedstock/pull/1
